### PR TITLE
ENC: Add parallel forward run

### PR DESCRIPTION
**Pull request by inoelloc**
Re establishing the parallel forward run for adjoint because Tapenade seems to be happy with it. The idea is that having re written the solver in the previous commit allows Tapenade to clarify the OMP atomic directive.

The parallel forward run is available for each module (snow, hydrological, routing). Snow and hydrological module are fully parallelized on the whole domain. For the routing module, new variables are calculated while generating the mesh (npar, ncpar, cscpar, cpar_to_rowcol and flwpar). This variables allow to parallelize the routing by independent partition.

Dataset meshs have been updated
Tests have been updated

### CODE REVIEW

File: smash/tests/core/simulation/test_bayesian_optimize.py [hunk 1]
```diff
@@ -10,7 +10,7 @@
 def generic_custom_bayesian_optimize(model: smash.Model, **kwargs) -> dict:
     res = {}
 
-    ncpu = max(1, os.cpu_count() - 1)
+    ncpu = min(5, os.cpu_count())
```
Comment (nghi-truyen) [inline smash/tests/core/simulation/test_bayesian_optimize.py:13]: Je propose ça pour éviter le fait d'utiliser tous les cpus pour les machines ayant que 1-5 cpus:
```suggestion
    ncpu = min(5, max(1, os.cpu_count() - 1))
```

File: smash/tests/core/simulation/test_estimate.py [hunk 2]
```diff
@@ -11,7 +11,7 @@
 def generic_multiset_estimate(model: smash.Model, **kwargs) -> dict:
     res = {}
 
-    ncpu = max(1, os.cpu_count() - 1)
+    ncpu = min(5, os.cpu_count())
```
Comment (nghi-truyen) [inline smash/tests/core/simulation/test_estimate.py:14]:
```suggestion
    ncpu = min(5, max(1, os.cpu_count() - 1))
```

File: smash/tests/core/simulation/test_optimize.py [hunk 3]
```diff
@@ -11,7 +11,7 @@
 def generic_optimize(model_structure: list[smash.Model], **kwargs) -> dict:
     res = {}
 
-    ncpu = max(1, os.cpu_count() - 1)
+    ncpu = min(5, os.cpu_count())
```
Comment (nghi-truyen) [inline smash/tests/core/simulation/test_optimize.py:14]:
```suggestion
    ncpu = min(5, max(1, os.cpu_count() - 1))
```

File: smash/tests/core/simulation/test_optimize.py [hunk 4]
```diff
@@ -92,7 +92,7 @@ def test_sparse_optimize():
 def generic_custom_optimize(model: smash.Model, **kwargs) -> dict:
     res = {}
 
-    ncpu = max(1, os.cpu_count() - 1)
+    ncpu = min(5, os.cpu_count())
```
Comment (nghi-truyen) [inline smash/tests/core/simulation/test_optimize.py:95]:
```suggestion
    ncpu = min(5, max(1, os.cpu_count() - 1))
```

File: smash/tests/core/simulation/test_optimize.py [hunk 5]
```diff
@@ -248,7 +248,7 @@ def test_custom_optimize():
 
 def test_multiple_optimize():
     instance = pytest.model.copy()
-    ncpu = max(1, os.cpu_count() - 1)
+    ncpu = min(5, os.cpu_count())
```
Comment (nghi-truyen) [inline smash/tests/core/simulation/test_optimize.py:251]:
```suggestion
    ncpu = min(5, max(1, os.cpu_count() - 1))
```

File: smash/tests/core/simulation/test_run.py [hunk 6]
```diff
@@ -10,7 +10,7 @@
 def generic_forward_run(model_structure: list[smash.Model], **kwargs) -> dict:
     res = {}
 
-    ncpu = max(1, os.cpu_count() - 1)
+    ncpu = min(5, os.cpu_count())
```
Comment (nghi-truyen) [inline smash/tests/core/simulation/test_run.py:13]:
```suggestion
    ncpu = min(5, max(1, os.cpu_count() - 1))
```

File: smash/tests/core/simulation/test_run.py [hunk 7]
```diff
@@ -74,7 +74,7 @@ def test_sparse_forward_run():
 
 def test_multiple_forward_run():
     instance = pytest.model.copy()
-    ncpu = max(1, os.cpu_count() - 1)
+    ncpu = min(5, os.cpu_count())
```
Comment (nghi-truyen) [inline smash/tests/core/simulation/test_run.py:77]:
```suggestion
    ncpu = min(5, max(1, os.cpu_count() - 1))
```

**Review by nghi-truyen — COMMENTED** [review]
Ok pour moi, juste une petite suggestion concernant le nombre de cpus qui vont utilisés pour les tests.

**Comment by inoelloc** [comment]
> Ok pour moi, juste une petite suggestion concernant le nombre de cpus qui vont utilisés pour les tests.

Done

**Review by nghi-truyen — APPROVED** [review]

**Pull request merged**